azure/services/securitygroups/securitygroups_test.go

```diff
@@ -18,212 +18,117 @@ package securitygroups
-					Name: to.StringPtr("nsg-one"),
-				}, nil)
-				m.CreateOrUpdate(gomockinternal.AContext(), "my-rg", "nsg-one", gomockinternal.DiffEq(network.SecurityGroup{
-					SecurityGroupPropertiesFormat: &network.SecurityGroupPropertiesFormat{
-						SecurityRules: &[]network.SecurityRule{
-							{
-								SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
-									Description:              to.StringPtr("a different rule"),
-									SourcePortRange:          to.StringPtr("*"),
-									DestinationPortRange:     to.StringPtr("*"),
-									SourceAddressPrefix:      to.StringPtr("*"),
-									DestinationAddressPrefix: to.StringPtr("*"),
-									Protocol:                 "*",
-									Direction:                "Outbound",
-									Access:                   "Deny",
-									Priority:                 to.Int32Ptr(300),
-								},
-								Name: to.StringPtr("foo-rule"),
-							},
-							{
-								SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
-									Description:              to.StringPtr("a test rule"),
-									SourcePortRange:          to.StringPtr("*"),
-									DestinationPortRange:     to.StringPtr("*"),
-									SourceAddressPrefix:      to.StringPtr("*"),
-									DestinationAddressPrefix: to.StringPtr("*"),
-									Protocol:                 "*",
-									Direction:                "Outbound",
-									Access:                   "Allow",
-									Priority:                 to.Int32Ptr(400),
-								},
-								Name: to.StringPtr("first-rule"),
-							},
-						},
-					},
-					Etag:     to.StringPtr("test-etag"),
-					Location: to.StringPtr("test-location"),
-				}))
-				m.Get(gomockinternal.AContext(), "my-rg", "nsg-two").Return(network.SecurityGroup{
-					Response: autorest.Response{},
-					SecurityGroupPropertiesFormat: &network.SecurityGroupPropertiesFormat{
-						SecurityRules: &[]network.SecurityRule{},
-					},
-					Etag: to.StringPtr("test-etag"),
-					ID:   to.StringPtr("fake/nsg/two/id"),
-					Name: to.StringPtr("nsg-two"),
-				}, nil)
+		},
+		{
+			name:          "first security groups create fails, should return error",
+			expectedError: errFake.Error(),
+			expect: func(s *mock_securitygroups.MockNSGScopeMockRecorder, r *mock_async.MockReconcilerMockRecorder) {
+				s.IsVnetManaged().Return(true)
+				s.NSGSpecs().Return([]azure.ResourceSpecGetter{&fakeNSG, &fakeNSG2})
+				r.CreateResource(gomockinternal.AContext(), &fakeNSG, serviceName).Return(nil, errFake)
+				r.CreateResource(gomockinternal.AContext(), &fakeNSG2, serviceName).Return(nil, nil)
+				s.UpdatePutStatus(infrav1.SecurityGroupsReadyCondition, serviceName, errFake)
 			},
-		}, {
-			name: "skipping network security group reconcile in custom VNet mode",
-			expect: func(s *mock_securitygroups.MockNSGScopeMockRecorder, m *mock_securitygroups.MockclientMockRecorder) {
+		},
+		{
+			name:          "first sg create fails, second sg create not done, should return create error",
```

Comment:
Just a thought: we have this block in several services where we have a loop and need to decide if we should overwrite the error message.
```
if !azure.IsOperationNotDoneError(err) || result == nil {
	result = err
}
```
We end up having to test the logic for handling `OperationNotDoneError`s and when `result == nil` for every service. Maybe we can try to refactor the logic to a helper that we can thoroughly test, similar to how we test all the errors related to `Get`, `CreateOrUpdateAsync`, and `Parameters` in `async_test.go`. For example, we might append all the errors to a list and create a function that picks the most important error out of it.